Automattic/wp-calypso
# InlineHelp: add admin sections

~This branch has been broken branched off from https://github.com/Automattic/wp-calypso/pull/43790. Let's review 43790 before to 🚀 this one.~

#### Changes proposed in this Pull Request

This PR adds admin sections to the results card component. It's rendered by the `<HelpSearch />` component in the customers home page, and almost the rest of the pages via the  `<InlineHelpPopover />` component (FAB)

### 🎨 Design

The following picture is the first approach for the design.

<img src="https://user-images.githubusercontent.com/4924246/86393393-dcbd7c80-bc51-11ea-9226-13140dd9e52a.gif" width="400px" /> | <img src="https://user-images.githubusercontent.com/4924246/86393619-2c9c4380-bc52-11ea-9af9-a4368075c9d6.png" width="300px" />
------------------------------

## About the technical implementation

### 💾 Static Data

The admin sections data is defined statically in the admin-sections.js file. It means if you want to add a new section or modify them, you will have to edit this file. The data structure follows the shape of the other ones, either the contextual help, defined in the contextual-help.js file, or the dynamic data which is collected from the endpoint response.
In order to identify these different data, this PR populates each item with the support_type key, defined by constants:

```es6
export const SUPPORT_TYPE_CONTEXTUAL_HELP = 'contextual_help';
export const SUPPORT_TYPE_API_HELP = 'api_help';
export const SUPPORT_TYPE_ADMIN_SECTION = 'admin_section';
```

### ⚙️ Processing Data

The changes purposed on this PR follows the current implementation of the inline-help state. Basically, all results data is organized by the `requestInlineHelpSearchResults()` action helper, in the state/inline-help/actions.js file. The idea here is to merge the dynamic data received from the endpoint response with the static data, which is the contextual help and the admin sections. In the end, the action passes all data as a single array to the reducer.
Remember that every result item has been populated with the `support_type` property, in order to differentiate it when it's required.

```es6
dispatch( {
	type: INLINE_HELP_SEARCH_REQUEST_SUCCESS,
	searchQuery,
	searchResults: hasAPIResults
	? [
		...mapWithSupportTypeProp( searchResults, SUPPORT_TYPE_API_HELP ),
		...helpAdminResults,
	  ]
	: [ ...contextualResults, ...helpAdminResults ],
} );
```

This is the only change that we do in data handling.

### 🖥️ UI

#### `<HelpSearchResults />`

This results card shows all items in the `<HelpSearch />` component (customer home page) or in the `<InlineHelpPopover />` (almost the rest of the other pages).
The results array gets by calling the `getSearchResultsByQuery()` action. It passes as a component property to finally to be rendered as results items.

The current item sets by calling the `selectSearchResult()` action. 

#### `<InlineHelpSearchCard />`

This component interacts with the `<HelpSearchResults />`. The following is a summary of the working flow:

1) `<InlineHelpSearchCard />`: The user types a term in the search field.
2) The app sets the query search through the `setInlineHelpSearchQuery()` action.
3) `<HelpSearchResults />`: The data query component _detects_ the search query changes because it's defined as one of its properties.
4) The data query component hits the API calling the `requestInlineHelpSearchResults()` action helper.
5) The helper processes the response and also populates it with static data (contextual help or admin sections ), triggering the `INLINE_HELP_SEARCH_REQUEST_SUCCESS` action.
6) The reducer populates the global state because of the `INLINE_HELP_SEARCH_REQUEST_SUCCESS` action.
7) The _selectedResult_ property refreshes producing a re-rendering in the `<HelpSearchResults />` component.

## 🧪 Testing instructions

* Apply these changes
* Go to customer home page `http://calypso.localhost:3000/home/<testing-site>` or open the InlineHelp popup clicking on the FAB icon
* Start to type a search term in the search card component. For instance, `media`

`<HelpSearch />` | `<InlineHelpPopover />`
------------------|------------------------
<img width="633" alt="Screen Shot 2020-07-07 at 10 19 28 AM" src="https://user-images.githubusercontent.com/77539/86787460-7cf11800-c03b-11ea-9361-c167fb92540e.png"> | <img width="347" alt="Screen Shot 2020-07-07 at 10 19 55 AM" src="https://user-images.githubusercontent.com/77539/86787447-782c6400-c03b-11ea-917b-fd854f3ffa67.png">

* Check the whole component, cleaning and retyping a new term.
* Navigate to listed sections.
* Use either the keyboard (up/down keys) as well by clicking to navigate.

### 🎈Addtionals 

#### Tracking events

* Check that it records a track action when an admin section is selected.
1) Set up debug to see recorded events: `localStorage.setItem( 'debug', 'calypso:analytics*' );`
2) Click on an admin section
3) Confirm it records it.

<img width="600" alt="Screen Shot 2020-07-07 at 10 30 41 AM" src="https://user-images.githubusercontent.com/77539/86788815-f50c0d80-c03c-11ea-933c-78d22f3f4fca.png">

#### Data persistence.

Confirm that either the search term as well as the search results data is persistent while the user navigates among app pages.

1) go to site homepage:  `http://calypso.localhost:3000/home/<testing-site>`
2) scroll to see the `Get Help...` card
3) Start to type a search term in the search card component. For instance, `podcast`
4) Take a look at the results shown by the component
5) Change the page, for instance going to the `stats`.
6) Open the InlineHelp popover by clicking on the Fab icon.
7) Confirm that it shows the same results as before.


Before | After
-------|-----
![image](https://user-images.githubusercontent.com/77539/86789364-9abf7c80-c03d-11ea-90b7-acbc2d7ad03c.png) | ![image](https://user-images.githubusercontent.com/77539/86789550-cfcbcf00-c03d-11ea-8b38-e33a3e94fb59.png)

## 📺 Test live

Use the [live branch](https://calypso.live/?branch=update/inline-help-add-admin-sections-ui) to test!
